extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java

```diff
@@ -307,6 +337,124 @@ private static void validateLimitAndOffset(final RelNode topRel, final boolean l
     }
   }
 
+  /**
+   * Validate that the query does not include any type changes from string to array or vice versa.
+   *
+   * These type changes tend to cause problems due to mixing of multi-value strings and string arrays. In particular,
+   * many queries written in the "classic MVD" style (treating MVDs as if they were regular strings) will fail when
+   * MVDs and arrays are mixed. So, we detect them as invalid.
+   *
+   * @param rootRel        root rel
+   * @param fieldMappings  field mappings from {@link #validateInsert(RelNode, List, Table, PlannerContext)}
+   * @param targetTable    table we are inserting (or replacing) into, if any
+   * @param plannerContext planner context
+   */
+  private static void validateTypeChanges(
+      final RelNode rootRel,
+      final List<Pair<Integer, String>> fieldMappings,
+      @Nullable final Table targetTable,
+      final PlannerContext plannerContext
+  )
+  {
+    if (targetTable == null) {
+      return;
```

Comment:
I suppose this doesn't necessarily need to be done as part of this PR, but I do want to see at some point that this would have a validation failure for the case where the output type has any ARRAY and `arrayIngestMode` isn't set to `array`, for the effect of preventing people from writing new insert queries that are incorrectly relying on the mvd behavior.

Reply:
I was thinking more about the validations that are in effect when `arrayIngestMode: array`, since those validations would help us change the default.

Maybe when we change the default, we can also consider making it so when people explicitly set `arrayIngestMode: mvd`, they need to use `ARRAY_TO_MV`. (I think that's what you're suggesting here.)

Reply:
>Maybe when we change the default, we can also consider making it so when people explicitly set arrayIngestMode: mvd, they need to use ARRAY_TO_MV. (I think that's what you're suggesting here.)

yea, I'm mostly thinking about it in terms of forcing people to write the correct queries as soon as possible in either mode to make things less painful down the road, and was thinking it would be nice to prevent bad habits for new tables before they have a chance to get started since those habits are going to have to change.

Its ok to wait until we change the default i guess, which this change should allow us to do much sooner (druid 30 hopefully)